### ISSUE TITLE
website and reality inbound with same port

### ISSUE THREAD

**Issue by DgithubA** [chika0801/Xray-examples#14]
hello. 
i have question about this.
i have some website in my vps which root path in /var/www/html was use ssl.(port 443)
now i want to use vless tls reality was use port 443 and www.google.com sni.
How do I configure nginx and xray to do this? that both my website works properly and I have inbound with port 443 using reverse port or anything

**Comment by Erfan-Fazeli** [comment]
https://github.com/chika0801/Xray-examples/blob/main/VLESS-HTTP2-REALITY/config_server_shared_port.json

https://github.com/chika0801/Xray-examples/blob/main/VLESS-Vision-REALITY/nginx_sni_shunting/nginx.conf

I think you can do it with these. I have been trying for 3 days, but because there is little documentation, I have not succeeded yet.
On the other hand, they do this with different sni
It means that the port and server are the same (for proxy and web server), but their domain is different and I don't know why they are not distinguished from each other by the /path
Is it impossible to separate the web server and proxy with /example? Because there is a document for tls, but I couldn't find anything for reality
@chika0801

**Comment by Erfan-Fazeli** [comment]
> now i want to use vless tls reality was use port 443 and [www.google.com](http://www.google.com) sni.

Why don't you use your own web server instead of Google?

**Comment by chika0801** [comment]
I personally don't recommend building a website and building a proxy, with the same VPS. because VPS is cheap, and you're short on that money and your time to learn how to 2-in-1.

Depending on your needs, you refer to this one link https://github.com/chika0801/Xray-examples/tree/main/VLESS-Vision-REALITY/nginx_sni_shunting

It's nginx listening on port 443, using a form of SNI triage, taking into account both the VLESS protocol and REALITY's steal yourself feature.

The nginx configuration file, you refer to and modify it yourself, it is an example of a reverse proxy in it.

**Comment by Erfan-Fazeli** [comment]
> I personally don't recommend building a website and building a proxy, with the same VPS. because VPS is cheap, and you're short on that money and your time to learn how to 2-in-1.

Unfortunately, reality has been recognized by MCI in Iran 
link : https://github.com/XTLS/Xray-core/issues/2778
My friends and I are conducting tests to determine how reality servers are detected and blocked Therefore, I plan to run a site and a reality server on a port 443 and use my site address for sni and dest, but according to your example, I could not establish a connection, but others were able to. No matter what I do, nginx gives me an error. Can you give me some more guidance?

**Comment by Erfan-Fazeli** [comment]
I plan not to use 2 domains and route it based on the route like the old fallback! Example: example.com/reality Because I want the address of the fake site and reality to be the same 
And don't routing based on sni Can you please put the sample in the notice of me and my friends? An example to be routed with the /example path

we need your help bro @chika0801

**Comment by chika0801** [comment]
> I plan not to use 2 domains and route it based on the route like the old fallback! Example: example.com/reality Because I want the address of the fake site and reality to be the same

What you are expressing is that you have a VPS that is listening on 443 using nginx and want to use the url path for triage.

For example www.123.com to access the content of the real website you built yourself with nginx
www.123.com/bbb passes the url path /bbb , to the VPS native Xray for proxying.

If you want to ask for such a form, with my knowledge, I don't think nginx has such a feature.

**Comment by chika0801** [comment]
The SNI triage feature of nginx that I know of just shares port 443 of the VPS.

nginx listens on port 443 to a different url 1.abc.com 2.abc.com, passes it to a different port locally 1.abc.com -> 127.0.0.1:8080 2.abc.com -> 127.0.0.1:9090.

So it can't achieve what you want 1.abc.com -> 127.0.0.1:8080 1.abc.com/bbb -> 127.0.0.1:9090

**Comment by chika0801** [comment]
https://github.com/XTLS/Xray-core/issues/2778

I can't help you with what is being discussed here.

**Comment by chika0801** [comment]
https://github.com/chika0801/Xray-examples/tree/main/VLESS-gRPC-TLS

Does your region block VLESS - gRPC - TLS by nginx listening for 443 to pass to backend Xray. this form?

For example, its variant VLESS/VMess - websocks - TLS is also used by nginx to listen for 443, which is passed to the backend Xray, and websocks has a URL meaning /path.

**Comment by chika0801** [comment]
VLESS/VMess - websocks - TLS

It's just that in mainland China, with this form, the VPS ports will be blocked by the firewall relatively quickly. Need to be able to change the port yourself. I don't recommend this combination.

It is used in China mainly by newcomers who watch old tutorials on youtube and are advertised to use CDNs, and newcomers who don't know how to use VPS for fear of having their IPs or ports blocked.

**Comment by Erfan-Fazeli** [comment]
In Iran reality is blocked quickly and sometimes in a few hours by mci. I was just doing a test to see if reality's port, address and sni belong to my site on the same server, will it also be blocked? So it is not possible in Iran. Should I use the path (/path) in port 443 for reality and reverse proxy and should the site be in the same port? And should I use the sni of the same site for my proxy server?
@chika0801

**Comment by chika0801** [comment]
> In Iran reality is blocked quickly and sometimes in a few hours by mci. I was just doing a test to see if reality's port, address and sni belong to my site on the same server, will it also be blocked


https://github.com/chika0801/Xray-examples/tree/main/VLESS-Vision-REALITY/steal_oneself

I don't really understand what you mean by that. In this way, do you use this combined form, with it, the SSL certificate has nginx loaded, for example, your SSL certificate contains the domain name 123.com, and you resolve this 123.com to the IP of your VPS in the DNS site.

And the Servername in the REALITY configuration is also filled with 123.com.

Have you tested using it this way?

**Comment by chika0801** [comment]
> And should I use the sni of the same site for my proxy server?

I built my own VPS and that's how I handled it, according to the principles above.

**Comment by TheErf23** [comment]
I intend to use both a reality server and a web server simultaneously on port 443. For instance, having a web server on port 80 and a reality server on port 1234. Nginx should listen on port 443, directing requests with "/" to the web server and "/reality" to the reality server. This allows me to test my theory regarding how MCI identifies reality servers. If my server gets blocked, it suggests a flaw in my theory. While the GitHub example you provided, nginx_sni_shunting, requires two domains, I need to use only one for my experiment. Currently, I've set up Nginx on my server on port 80 and a reality proxy on port 1234, configured Nginx for SSL on port 443, and want specific paths to be directed to the reality server. This approach aims to ensure that the server's IP, domain, and port for the web server and reality match, helping verify my theory despite its limitations with non-reality services like CDNs. Is it possible to route requests with "/" to the web server and "/reality" to the reality server based on these details?

**Comment by Erfan-Fazeli** [comment]
I believe using Google Translate may have caused some misunderstanding. I want to use a reality server and a web server simultaneously on port 443. For instance, having a web server on port 80 and a reality server on port 1234. Nginx should listen on port 443, sending requests with "/" to the web server and "/reality" to the reality server. This helps test my theory on how MCI identifies reality servers. If my server gets blocked again, it indicates a flaw in my theory. The example you shared on GitHub, "nginx_sni_shunting," is similar, but it requires two domains, while I need to use only one for my experiment.

Currently, I've set up Nginx on my server on port 80 and a reality proxy on port 1234. I've configured Nginx, obtained SSL for my domain, and set up reverse proxy from port 443 to port 80 where my web server is. Now, my site is HTTPS with valid SSL.

Now, I want Nginx to send requests with a specific path like abc.com/reality to port 12345. In the SNI and dest sections, I want to use my web server and website address for reality.

The goal of this is to use a simple method for server identification, considering the limitations of the theory, especially with non-reality services like CDNs. If the Iranian government's GFW suspects a server, it checks the IP behind SNI. If the IP of the site doesn't match your reality server's IP, it gets blocked. Although this theory has flaws, as it affects services like CDNs, I want to ensure that the server's IP, domain, and port for the web server and reality are exactly the same. If my server gets blocked by MCI, it means my theory is incorrect. With these explanations, is it possible to direct requests with "/" to the web server and "/reality" to the reality server?

**Comment by Erfan-Fazeli** [comment]
> https://github.com/chika0801/Xray-examples/tree/main/VLESS-Vision-REALITY/steal_oneself

What is this sample and what does it do? Why had I not seen him😐

**Comment by Erfan-Fazeli** [comment]
I still doubt that Google Translate has correctly expressed my meaning to you, so I will try to tell you what I mean with some pictures. 

First, I installed an aapanel to be able to manage the site more easily, and the photos belong to aapanel and nothing special (consider an extra step)


![Screenshot 2023-12-05 212314](https://github.com/chika0801/Xray-examples/assets/82929677/3b6cce41-8563-44ac-8901-58b4394dc927)
The web server and site are running on port 80

![Screenshot 2023-12-05 212406](https://github.com/chika0801/Xray-examples/assets/82929677/572e24cd-a6f1-4bb0-aadc-fb8982daa838)
![Screenshot 2023-12-05 213019](https://github.com/chika0801/Xray-examples/assets/82929677/bb74b289-80e7-4dde-b862-eda2c2446058)


and SSL is valid





![image](https://github.com/chika0801/Xray-examples/assets/82929677/a556ac66-0984-4f9c-84d2-7354825cc51c)


my reality server (tcp or h2) on 1234 port



```
server
{
    listen 80;
		listen 443 ssl http2;
		listen [::]:443 ssl http2;
    listen [::]:80;
    server_name mysite.com;
    index index.php index.html index.htm default.php default.htm default.html;
    root /www/wwwroot/mysite.com;

    #SSL-START SSL related configuration, do NOT delete or modify the next line of commented-out 404 rules
    #error_page 404/404.html;
    ssl_certificate    /www/server/panel/vhost/cert/mysite.com/fullchain.pem;
    ssl_certificate_key    /www/server/panel/vhost/cert/mysite.com/privkey.pem;
    ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;
    ssl_ciphers EECDH+CHACHA20:EECDH+CHACHA20-draft:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:!MD5;
    ssl_prefer_server_ciphers on;
    ssl_session_cache shared:SSL:10m;
    ssl_session_timeout 10m;
    add_header Strict-Transport-Security "max-age=31536000";
    error_page 497  https://$host$request_uri;
		#SSL-END

    #ERROR-PAGE-START  Error page configuration, allowed to be commented, deleted or modified
    #error_page 404 /404.html;
    #error_page 502 /502.html;
    #ERROR-PAGE-END

    #PHP-INFO-START  PHP reference configuration, allowed to be commented, deleted or modified
    include enable-php-80.conf;
    #PHP-INFO-END

    #REWRITE-START URL rewrite rule reference, any modification will invalidate the rewrite rules set by the panel
    include /www/server/panel/vhost/rewrite/mysite.com.conf;
    #REWRITE-END

    # Forbidden files or directories
    location ~ ^/(\.user.ini|\.htaccess|\.git|\.env|\.svn|\.project|LICENSE|README.md)
    {
        return 404;
    }

    # Directory verification related settings for one-click application for SSL certificate
    location ~ \.well-known{
        allow all;
    }

    #Prohibit putting sensitive files in certificate verification directory
    if ( $uri ~ "^/\.well-known/.*\.(php|jsp|py|js|css|lua|ts|go|zip|tar\.gz|rar|7z|sql|bak)$" ) {
        return 403;
    }

    location ~ .*\.(gif|jpg|jpeg|png|bmp|swf)$
    {
        expires      30d;
        error_log /dev/null;
        access_log off;
    }

    location ~ .*\.(js|css)?$
    {
        expires      12h;
        error_log /dev/null;
        access_log off; 
    }
    access_log  /www/wwwlogs/mysite.com.log;
    error_log  /www/wwwlogs/mysite.com.error.log;
}
```

and my nginx config
The web server listens to 443

I want if the input url
mysite.com was the site and if the input url
mysite.com/reality was to go to port 1234 and reality server


Something like this that I have already done with TLS with the help of Fallback and Nginx:

![image](https://github.com/chika0801/Xray-examples/assets/82929677/e08de9ef-1ceb-4f1d-abf2-a42505a75323)

In the sample photo above, with the help of the available samples, I have done the same thing with the help of tls, but I don't know if it can be done with Reality or not.
Something like this, I want to use a domain and route with url and path /eample :)

**Comment by chika0801** [comment]
> Now, I want Nginx to send requests with a specific path like abc.com/reality to port 12345. In the SNI and dest sections, I want to use my web server and website address for reality.

It's out of my league, and I've thought about REALITY, and I don't think I can achieve that. My earlier reply to you also meant it couldn't.

> mysite.com/reality was to go to port 1234 and reality server

It can't.

https://github.com/chika0801/Xray-examples/issues/14#issuecomment-1842056239

How about you try this combo form.

**Comment by Erfan-Fazeli** [comment]
> How about you try this combo form.

I will try because it is my last hope. Can you explain a little about it and tell me how it works because I don't understand
Did he really upload this sample 5 days ago? I visit your GitHub every day, I really don't know how I didn't see it
Anyway, I think it solves my problem
I have another question in terms of security, which one is safer and better in your opinion?
steal_oneself or nginx_sni_shunting?
Thank you, Chika

**Comment by chika0801** [comment]
https://github.com/chika0801/Xray-examples/tree/main/VLESS-Vision-REALITY/steal_oneself

This example is listening on port 443 by Xray, with the address 127.0.0.1:8001 filled in by dest, and listening on port 127.0.0.1:8001 by nginx, with the SSL certificate loaded by nginx.

Normal requests to access the website, which do not meet the requirements of the REALITY parameter, are redirected to 127.0.0.1:8001, where nginx serves the content (in the example, a website is reverse proxied).

The point of attention is SNI, as I told you earlier, for example, your SSL certificate contains the domain name 123.com, you point it correctly to the IP of the current VPS, you enter 123.com in the browser will display the content of the nginx reverse proxy website normally.

And the servername in the Xray configuration is also filled with 123.com, the SNI of the request from your client is 123.com, and the SNI of your request as seen by your operator's firewall is also 123.com.

**Comment by TheErf23** [comment]
> This example is listening on port 443 by Xray, with the address 127.0.0.1:8001 filled in by dest, and listening on port 127.0.0.1:8001 by nginx, with the SSL certificate loaded by nginx.

I think this is great
And thank you, tell me which one is more safe between the two samples and I have no more questions😂
steal_oneself or nginx_sni_shunting?
If you were a gfw yourself, which one would you be less suspicious of?😂

**Comment by DgithubA** [comment]
https://github.com/hiddify/Hiddify-Manager
see this repository. It has done what I want using [haproxy](https://github.com/haproxy/haproxy/).
but i want use x-ui with more than custumize like add header,etc.

**Comment by DgithubA** [comment]
i thinks [my friend](@TheErf23) say all thinks i want.
thanks for your help.
its better myself to read xray,reality and nginx documentation and improve network knowledge to pass GFW of iran.

**Comment by DgithubA** [comment]
@TheErf23 
bro can give me some telegram id. i want to contact u about this discusse.
(or send message to my telegram in my bio)
